In [56]:
import pandas as pd

tabela = pd.read_csv("cancelamentos.csv")
display(tabela)

CustomerID  idade    sexo  tempo_como_cliente  frequencia_uso  \
0        349936.0   23.0    Male                13.0            22.0   
1        100634.0   49.0    Male                55.0            16.0   
2        301263.0   30.0    Male                 7.0             1.0   
3        119358.0   26.0    Male                40.0             5.0   
4        130955.0   27.0  Female                17.0            30.0   
...           ...    ...     ...                 ...             ...   
49995    195680.0   62.0  Female                35.0             7.0   
49996     43477.0   36.0    Male                43.0            21.0   
49997    169273.0   55.0    Male                42.0             8.0   
49998    310693.0   40.0  Female                14.0            19.0   
49999    251915.0   64.0    Male                41.0            29.0   

       ligacoes_callcenter  dias_atraso assinatura duracao_contrato  \
0                      2.0          1.0   Standard           Annual   
1                      3.0          6.0    Premium          Monthly   
2                      0.0          8.0      Basic           Annual   
3                      3.0          8.0    Premium           Annual   
4                      5.0          6.0      Basic           Annual   
...                    ...          ...        ...              ...   
49995                  2.0          8.0      Basic           Annual   
49996                  2.0         30.0      Basic        Quarterly   
49997                  1.0         12.0      Basic          Monthly   
49998                  1.0         17.0    Premium        Quarterly   
49999                  5.0          9.0      Basic        Quarterly   

       total_gasto  meses_ultima_interacao  cancelou  
0           909.58                    23.0       0.0  
1           207.00                    29.0       1.0  
2           768.78                     7.0       0.0  
3           398.00                    12.0       1.0  
4           507.00                    15.0       1.0  
...            ...                     ...       ...  
49995       232.00                    15.0       1.0  
49996       928.00                    30.0       1.0  
49997       326.00                    27.0       1.0  
49998       826.76                    12.0       0.0  
49999       700.38                    18.0       1.0  

[50000 rows x 12 columns]

## Tratando os Dados
Precisamos fazer uma limpeza na tabela para que não tenhamos dados em excesso.   
A coluna CustomerID não será relevante para nossa análise de dados, por isso vamos apagar

In [57]:
tabela = tabela.drop("CustomerID", axis=1)
display(tabela)

idade    sexo  tempo_como_cliente  frequencia_uso  ligacoes_callcenter  \
0       23.0    Male                13.0            22.0                  2.0   
1       49.0    Male                55.0            16.0                  3.0   
2       30.0    Male                 7.0             1.0                  0.0   
3       26.0    Male                40.0             5.0                  3.0   
4       27.0  Female                17.0            30.0                  5.0   
...      ...     ...                 ...             ...                  ...   
49995   62.0  Female                35.0             7.0                  2.0   
49996   36.0    Male                43.0            21.0                  2.0   
49997   55.0    Male                42.0             8.0                  1.0   
49998   40.0  Female                14.0            19.0                  1.0   
49999   64.0    Male                41.0            29.0                  5.0   

       dias_atraso assinatura duracao_contrato  total_gasto  \
0              1.0   Standard           Annual       909.58   
1              6.0    Premium          Monthly       207.00   
2              8.0      Basic           Annual       768.78   
3              8.0    Premium           Annual       398.00   
4              6.0      Basic           Annual       507.00   
...            ...        ...              ...          ...   
49995          8.0      Basic           Annual       232.00   
49996         30.0      Basic        Quarterly       928.00   
49997         12.0      Basic          Monthly       326.00   
49998         17.0    Premium        Quarterly       826.76   
49999          9.0      Basic        Quarterly       700.38   

       meses_ultima_interacao  cancelou  
0                        23.0       0.0  
1                        29.0       1.0  
2                         7.0       0.0  
3                        12.0       1.0  
4                        15.0       1.0  
...                       ...       ...  
49995                    15.0       1.0  
49996                    30.0       1.0  
49997                    27.0       1.0  
49998                    12.0       0.0  
49999                    18.0       1.0  

[50000 rows x 11 columns]

In [60]:
display(tabela.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49996 entries, 0 to 49999
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   idade                   49996 non-null  float64
 1   sexo                    49996 non-null  object 
 2   tempo_como_cliente      49996 non-null  float64
 3   frequencia_uso          49996 non-null  float64
 4   ligacoes_callcenter     49996 non-null  float64
 5   dias_atraso             49996 non-null  float64
 6   assinatura              49996 non-null  object 
 7   duracao_contrato        49996 non-null  object 
 8   total_gasto             49996 non-null  float64
 9   meses_ultima_interacao  49996 non-null  float64
 10  cancelou                49996 non-null  float64
dtypes: float64(8), object(3)
memory usage: 4.6+ MB


None

Ao exibir as informações da tabela `display(tabela.info())` podemos observar que nem todas as linhas possuem os dados completos. 
Vamos utilizar o comando `dropna` para remover as informações vazias da tabela

In [61]:
tabela = tabela.dropna()
display(tabela.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49996 entries, 0 to 49999
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   idade                   49996 non-null  float64
 1   sexo                    49996 non-null  object 
 2   tempo_como_cliente      49996 non-null  float64
 3   frequencia_uso          49996 non-null  float64
 4   ligacoes_callcenter     49996 non-null  float64
 5   dias_atraso             49996 non-null  float64
 6   assinatura              49996 non-null  object 
 7   duracao_contrato        49996 non-null  object 
 8   total_gasto             49996 non-null  float64
 9   meses_ultima_interacao  49996 non-null  float64
 10  cancelou                49996 non-null  float64
dtypes: float64(8), object(3)
memory usage: 4.6+ MB


None

## Análise da taxa de cancelamento
Com a base de dados tratada, iniciaremos a análise verificando a taxa de cancelamento, tendo em vista que o objetivo é diminuir essa taxa

In [62]:
# Verificando a quantidade de pessoas que cancelaram
display(tabela["cancelou"].value_counts())

1.0    28393
0.0    21603
Name: cancelou, dtype: int64

In [63]:
# Colocando em porcentagem
display(tabela["cancelou"].value_counts(normalize=True))

1.0    0.567905
0.0    0.432095
Name: cancelou, dtype: float64

In [64]:
# Alterando visualização
display(tabela["cancelou"].value_counts(normalize=True).map("{:.1%}".format))

1.0    56.8%
0.0    43.2%
Name: cancelou, dtype: object

## Verificando a Duração do Contrato

In [65]:
display(tabela["duracao_contrato"].value_counts())

Annual       20156
Quarterly    19956
Monthly       9884
Name: duracao_contrato, dtype: int64

In [66]:
display(tabela["duracao_contrato"].value_counts(normalize=True))

Annual       0.403152
Quarterly    0.399152
Monthly      0.197696
Name: duracao_contrato, dtype: float64

## Analisando as Informações dos Contratos 

In [67]:
# analisando contrato mensal
display(tabela.groupby("duracao_contrato").mean(numeric_only=True))

idade  tempo_como_cliente  frequencia_uso  \
duracao_contrato                                                  
Annual            38.783985           31.416452       15.910449   
Monthly           41.428470           30.677964       15.577600   
Quarterly         38.833935           31.522099       15.842504   

                  ligacoes_callcenter  dias_atraso  total_gasto  \
duracao_contrato                                                  
Annual                       3.277585    12.533985   650.925840   
Monthly                      4.923917    15.078814   547.508921   
Quarterly                    3.256113    12.480006   654.102443   

                  meses_ultima_interacao  cancelou  
duracao_contrato                                    
Annual                         14.231544  0.464080  
Monthly                        15.392655  1.000000  
Quarterly                      14.364602  0.458759

Com as informações agrupadas, é possível notar que os clientes do plano Mensal, possuem uma média de cancelamento igual a 1, ou seja, praticamente todos os clientes que utilizam esse plano fizeram o cancelamento do serviço.

## Removendo o contrato mensal
Sabendo que o contrato mensal é ruim para a empresa, podemos remover as informações desse contrato específico e continuar analisando.

In [68]:
tabela = tabela[tabela["duracao_contrato"]!="Monthly"]
display(tabela)

idade    sexo  tempo_como_cliente  frequencia_uso  ligacoes_callcenter  \
0       23.0    Male                13.0            22.0                  2.0   
2       30.0    Male                 7.0             1.0                  0.0   
3       26.0    Male                40.0             5.0                  3.0   
4       27.0  Female                17.0            30.0                  5.0   
6       49.0    Male                 6.0             7.0                  0.0   
...      ...     ...                 ...             ...                  ...   
49994   63.0    Male                16.0            24.0                  2.0   
49995   62.0  Female                35.0             7.0                  2.0   
49996   36.0    Male                43.0            21.0                  2.0   
49998   40.0  Female                14.0            19.0                  1.0   
49999   64.0    Male                41.0            29.0                  5.0   

       dias_atraso assinatura duracao_contrato  total_gasto  \
0              1.0   Standard           Annual       909.58   
2              8.0      Basic           Annual       768.78   
3              8.0    Premium           Annual       398.00   
4              6.0      Basic           Annual       507.00   
6              0.0   Standard           Annual       751.00   
...            ...        ...              ...          ...   
49994         18.0   Standard        Quarterly       442.00   
49995          8.0      Basic           Annual       232.00   
49996         30.0      Basic        Quarterly       928.00   
49998         17.0    Premium        Quarterly       826.76   
49999          9.0      Basic        Quarterly       700.38   

       meses_ultima_interacao  cancelou  
0                        23.0       0.0  
2                         7.0       0.0  
3                        12.0       1.0  
4                        15.0       1.0  
6                        11.0       0.0  
...                       ...       ...  
49994                    26.0       1.0  
49995                    15.0       1.0  
49996                    30.0       1.0  
49998                    12.0       0.0  
49999                    18.0       1.0  

[40112 rows x 11 columns]

In [69]:
display(tabela["cancelou"].value_counts())
display(tabela["cancelou"].value_counts(normalize=True).map("{:.1%}".format))

0.0    21603
1.0    18509
Name: cancelou, dtype: int64

0.0    53.9%
1.0    46.1%
Name: cancelou, dtype: object

Após remover os assinantes mensais, podemos observar que a taxa de cancelamento caiu para 46.1%, porém, ainda precisamos avançar na análise

## Análise de Assinaturas

In [70]:
display(tabela["assinatura"].value_counts(normalize=True))
display(tabela.groupby("assinatura").mean(numeric_only=True))

Standard    0.343264
Premium     0.333466
Basic       0.323270
Name: assinatura, dtype: float64

idade  tempo_como_cliente  frequencia_uso  \
assinatura                                                  
Basic       38.777049           32.266908       15.938459   
Premium     38.885766           31.120813       15.762111   
Standard    38.764035           31.055850       15.929697   

            ligacoes_callcenter  dias_atraso  total_gasto  \
assinatura                                                  
Basic                  3.328064    12.538752   648.339261   
Premium                3.238038    12.430697   653.459842   
Standard               3.237345    12.551601   655.504070   

            meses_ultima_interacao  cancelou  
assinatura                                    
Basic                    14.317036  0.476517  
Premium                  14.306071  0.453947  
Standard                 14.271479  0.454499

Podemos observar que a quantidade de assinantes de Standard / premium / basic é bem parecida e que a taxa de cancelamento também é parecida.

## Análises Gráficas
Como a última análise não foi muito boa vamos criar gráficos porque só com números está difícil visualizar

In [76]:
import plotly.express as px

grafico = px.histogram(tabela, x=tabela["ligacoes_callcenter"], color = "cancelou", width=600)
#grafico.show()

grafico = px.histogram(tabela, x=tabela["dias_atraso"], color = "cancelou", width=600)
#grafico.show()

#for coluna in tabela.columns:
    #grafico = px.histogram(tabela, x=coluna, color="cancelou", width=600)
    #grafico.show()

In [72]:
tabela = tabela[tabela["ligacoes_callcenter"]<5]
tabela = tabela[tabela["dias_atraso"]<=20]
display(tabela)

idade    sexo  tempo_como_cliente  frequencia_uso  ligacoes_callcenter  \
0       23.0    Male                13.0            22.0                  2.0   
2       30.0    Male                 7.0             1.0                  0.0   
3       26.0    Male                40.0             5.0                  3.0   
6       49.0    Male                 6.0             7.0                  0.0   
7       33.0    Male                15.0            18.0                  1.0   
...      ...     ...                 ...             ...                  ...   
49991   24.0  Female                33.0            28.0                  1.0   
49992   37.0    Male                46.0            29.0                  2.0   
49994   63.0    Male                16.0            24.0                  2.0   
49995   62.0  Female                35.0             7.0                  2.0   
49998   40.0  Female                14.0            19.0                  1.0   

       dias_atraso assinatura duracao_contrato  total_gasto  \
0              1.0   Standard           Annual       909.58   
2              8.0      Basic           Annual       768.78   
3              8.0    Premium           Annual       398.00   
6              0.0   Standard           Annual       751.00   
7             20.0   Standard           Annual       839.71   
...            ...        ...              ...          ...   
49991         11.0    Premium           Annual       844.05   
49992          9.0   Standard        Quarterly       517.86   
49994         18.0   Standard        Quarterly       442.00   
49995          8.0      Basic           Annual       232.00   
49998         17.0    Premium        Quarterly       826.76   

       meses_ultima_interacao  cancelou  
0                        23.0       0.0  
2                         7.0       0.0  
3                        12.0       1.0  
6                        11.0       0.0  
7                        25.0       0.0  
...                       ...       ...  
49991                     9.0       0.0  
49992                     3.0       0.0  
49994                    26.0       1.0  
49995                    15.0       1.0  
49998                    12.0       0.0  

[26267 rows x 11 columns]

In [73]:
display(tabela["cancelou"].value_counts())
display(tabela["cancelou"].value_counts(normalize=True).map("{:.1%}".format))

0.0    21446
1.0     4821
Name: cancelou, dtype: int64

0.0    81.6%
1.0    18.4%
Name: cancelou, dtype: object

Com os gráficos descobrimos que acima de 20 dias de atrasim cancelam e ligações call center acima de 5 todos cancelam.
Se resolvermos isso, a taxa de cancelamento cai para 18.4%